### PR TITLE
Crate chain-pushing try 2

### PR DESCRIPTION
**Pull request by Dennok**
## About The Pull Request

Thanks @XDTM for initial PR.
I take #27510 and fix one string.
99% runtimes less.
Conveyor tested.

## Why It's Good For The Game

If you're pushing a crate and there's a crate in front of that crate, you'll push that crate as well. Pretty much QoL for cargo, so they don't have to push every single crate one by one to get them on/off the shuttle.

## Changelog
:cl:
add: You can now push multiple crates in front of you. This also applies to conveyor belts.
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
Atleast make you move slower if you're pushing like 5 crates.

**Comment by Dennok** [comment]
I dont make it ideal clear chain pushing, you need 3 steps to move 5 crates on one turf. 
When you push crates, move only 2 last.

**Comment by MrDoomBringer** [comment]
ree quickly loading the shuttle by dragging one crate and pushing another infront of you is an art and should not be disrespected

**Comment by Ziiro** [comment]
the future is now old man

**Comment by optimumtact** [comment]
You need to provide document comments on those class variables as per
https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md

**Comment by Bawhoppen** [comment]
But now cargo won't be a Sokoban minigame anymore

**Comment by Dennok** [comment]
> ree quickly loading the shuttle by dragging one crate and pushing another infront of you is an art and should not be disrespected

Now you can pull crate while push 5 crates.

**Comment by Tharcoonvagh** [comment]
F I N A L L Y

**Comment by nemvar** [comment]
> ree quickly loading the shuttle by dragging one crate and pushing another infront of you is an art and should not be disrespected

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 1]
```diff
@@ -20,6 +23,43 @@
...
+		///hit an obstacle, previous crate must not try to move or it'll loop
+		bump_chain = 0
+	else
+		var/obj/structure/closet/crate/C = A
+		if(C.anchored)
+			///an anchored crate is an obstacle
+			bump_chain = 0
+
+/// When crate bumped by something called Bumped(AM)
+/obj/structure/closet/crate/Bumped(AM as mob|obj)
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]:
```suggestion
/obj/structure/closet/crate/Bumped(atom/movable/AM)

```

**Review by spookydonut — CHANGES_REQUESTED** [review]

**Comment by Dennok** [comment]
While i think about this i got idea make "Cargo" skill that allow you push more crates.

**Comment by XDTM** [comment]
i initially made the feature to let conveyor belts push a whole line of crates into the shuttle, so placing it into a skill would defeat the point

**Comment by Dennok** [comment]
> i initially made the feature to let conveyor belts push a whole line of crates into the shuttle, so placing it into a skill would defeat the point

OMG, you just need to make belts better.
With this belts now can lose crates on conweyor turn. Now on turns need place blocker (anchored crate or glass is nice), because back crates can push front crates out of conweyor if you owerload it.

**Comment by XDTM** [comment]
that was unintended, but you could potentially want that sometimes

```
>>>>vvvv
    vvvv
    vvvv
    vvvv
```

that structure would let the crates fill the other rows once the first one is full, which would be ideal if you wanted to fill the shuttle without having to manually move a single crate

**Comment by Spepbuthewranglesgithub** [comment]
ah yes, heruclownes.

**Comment by Bawhoppen** [comment]
I am really actually not a fan of this, for a few reasons:
While it may have been a meme, ss13 is founded on having a bunch of minigames all encompassed within it... Having cargo be Sokoban-esque fits well with that trend, and I would hate to see that go.

More importantly, this is one of the situations where we'd be removing some of the character from the game. Sure, it's making things more streamlined, and efficient (all for the sake of QoL), but at the same time, we're losing a little bit of the unique charm that makes this game so great. Not everything needs to be maximally efficient or hyper-sleek; there's sometimes a benefit to preserving neat little quirks like this, even if it means things will be slightly more clunky.

**Comment by XDTM** [comment]
I think there's people that find joy by doing jobs and tasks, and others who instead like putting effort in making something that lets them avoid that job, factorio style. I suppose this feature is the kind of thing that caters to the latter at the expense of the former.

**Comment by MrDoomBringer** [comment]
Another issue is cargo is a job that we tell new players to pick - only being able to push one object at a time is an important part of learning how to maneuver around in this game - for example getting stuck in maint tunnels when there's two dense objects infront of you, etc. I get that this is a nitpicky comment, but still.

**Comment by cacogen** [comment]
> **OOC:  Bawhoppennn :  hey people 
> OOC:  Bawhoppennn :  do you all trust my judgment 
> OOC:  Bawhoppennn :  I need you all to downvote a PR 
> OOC:  Cacogen :  absolutely 
> OOC:  Cacogen :  lay it on me 
> OOC:  MarcusRoland:  What pr 
> OOC:  Kendrickorium:  i dopwnvote all prs 
> OOC:  Bawhoppennn :  downvote this PR please: https://github.com/tgstation/tgstation/pull/49657  
> OOC:  Bawhoppennn :  just trust me 
> OOC:  Bawhoppennn :  I know it may seem good at face value but it's not 
> OOC:  Lithra:  That's... a badly balanced PR 
> OOC:  Cacogen :  how can i when i already clicked the upward facing thumb, the rocket ship and the love heart 
> OOC:  Zytolg :  That's a horrible PR 
> OOC:  Bawhoppennn :  yes please give it a downvote 
> OOC:  MarcusRoland:  Done 
> OOC:  TheMidnightRose:  Blegh 
> OOC:  Moist Cookie :  upvoted 
> OOC:  Lithra:  At most maybe 2 crates with a stamina hit. 
> OOC:  Cacogen :  do not brigade dennok's PR 
> OOC:  Jack7D1 :  upvoted 
> OOC:  Bawhoppennn :  I will do everything in my power to stop this PR 
> OOC:  Bawhoppennn :  it will ruin the classic fun of crate managing in cargo 
> OOC:  Lithra:  Also ">Qol for cargo" They should use a fucking conveyor system with the hand tele. 
> OOC:  Zytolg :  It really shouldn't be any crates at all 
> OOC:  Bawhoppennn :  QoL is buzzword for making game more simplified 
> OOC:  Lithra:  Basically Bawhooppenn. Cargo has access to things to MAKE stuff to make it easier. 
> OOC:  Zytolg :  The floors are tiled, so if you're pusing more than one crate, chances are it'd still get stuck and snag on a crack 
> OOC:  Bawhoppennn :  yeah that is my thought also 
> OOC:  Lithra:  Cargo already doesn't have much to do. 
> OOC:  Zytolg :  That's my headcannon for it anyways 
> OOC:  Bawhoppennn :  then it removes the potential future of them having like crate trailers and stuff 
> OOC:  Lithra:  Yes, I'd love that honestly. 
> OOC:  Zytolg :  Cargo has a shit ton today 
> OOC:  Bawhoppennn :  please comment about it 
> OOC:  Lithra:  Cargo would be 10/10 with their own rideable vehicle. 
> OOC:  Zytolg :  people just don't often chase those dreams and ambitions 
> OOC:  Bawhoppennn :  we need an upswell of grassoots disapproval 
> OOC:  Zytolg :  Cargo needs a golf cart 
> OOC:  Bawhoppennn :  WE NEED THE GRASSROOTS SUPPORT 
> OOC:  Lithra:  Nah cargo can get ATV's 
> OOC:  Bawhoppennn :  I've already posted 2 comments I can't afford another one 
> OOC:  Moist Cookie :  the OG ripley is good for cargo, you know 
> OOC:  Bawhoppennn :  it is true 
> OOC:  Moist Cookie :  you can kidnap so much stuff in it 
> OOC:  Bawhoppennn :  3:1 was a good ratio thanks OOC**

**Comment by Bawhoppen** [comment]
Bawhoppen was simply spreading awareness of a PR they were gravely concerned about, and expressed their opinion about it... No individual was under any obligation to follow their suggestion!!!

:woman:PEOPLE:man:  :-1: DOWNVOTED:-1: OF THEIR :moneybag:OWN:moneybag: VOLITION!

**Comment by cacogen** [comment]
you know i had to do it to em

**Comment by Dennok** [comment]
This DONT "removes the potential future of them having like crate trailers and stuff"
You can push crates only straight. And crates now anyway not unpassable things, you can climb on crates. And crate can be anchored to floor if you vant unmovable crate.

**Comment by XDTM** [comment]
i'm really disappointed in the downvote brigading

there's proper reasons to not want this, but 
>OOC: Bawhoppennn : I know it may seem good at face value but it's not

is not really one of them

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 2]
```diff
@@ -20,6 +23,44 @@
 	drag_slowdown = 0
 	var/obj/item/paper/fluff/jobs/cargo/manifest/manifest
 
+	///counter for crates that push this one. Start from 1 and increment when Bumped.
+	var/bump_chain = 1
+
+/// When crate fail move to another place it call Bump(atom/A)
+/obj/structure/closet/crate/Bump(atom/A)
+	..()
+	///check that the crate can push the obstacle that is blocking it or not.
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: you dont need to add comments like this, this isnt what the code docs requirement is

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 3]
```diff
@@ -20,6 +23,44 @@
...
 
+	///counter for crates that push this one. Start from 1 and increment when Bumped.
+	var/bump_chain = 1
+
+/// When crate fail move to another place it call Bump(atom/A)
+/obj/structure/closet/crate/Bump(atom/A)
+	..()
+	///check that the crate can push the obstacle that is blocking it or not.
+	if(!istype(A, /obj/structure/closet/crate))
+		///hit an obstacle, previous crate must not try to move or it'll loop
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 4]
```diff
@@ -20,6 +23,44 @@
...
+		bump_chain = 0
+	else
+		var/obj/structure/closet/crate/C = A
+		if(C.anchored)
+			///an anchored crate is an obstacle
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 5]
```diff
@@ -20,6 +23,44 @@
...
+			bump_chain = 0
+
+/// When crate bumped by something called Bumped(AM)
+/obj/structure/closet/crate/Bumped(atom/movable/AM)
+
+	..()
+	///check, that crate bumbed by crate and can move.
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 6]
```diff
@@ -20,6 +23,44 @@
...
+	if(istype(AM, /obj/structure/closet/crate) && !anchored)
+		var/obj/structure/closet/crate/previous_crate = AM
+		bump_chain = previous_crate.bump_chain + 1
+		///check for limit of bumped chain.
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 7]
```diff
@@ -20,6 +23,44 @@
...
+		if(bump_chain > MAX_CRATE_BUMP_CHAIN)
+			///when reached, reset self and do nothing.
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 8]
```diff
@@ -20,6 +23,44 @@
...
+			bump_chain = 1
+			return
+		var/d = get_dir(AM, src)
+		
+		/// Try to move next crate ,did it hit an obstacle?
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 9]
```diff
@@ -20,6 +23,44 @@
...
+		if(step(src, d))
+			///move to follow
+			step(AM, d)
+	///reset of chain counter
+	bump_chain = 1
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: same

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 10]
```diff
@@ -20,6 +23,44 @@
 	drag_slowdown = 0
 	var/obj/item/paper/fluff/jobs/cargo/manifest/manifest
 
+	///counter for crates that push this one. Start from 1 and increment when Bumped.
+	var/bump_chain = 1
+
+/// When crate fail move to another place it call Bump(atom/A)
```
Comment (spookydonut) [inline code/game/objects/structures/crates_lockers/crates.dm]: you dont need this kind of comment either, Bump() is a byond built-in

**Review by spookydonut — COMMENTED** [review]
The other comments are good and satisfy the code docs requirement.

You dont need to remove any of these, this is just as an FYI you don't need to go overboard like this.

**Review by spookydonut — APPROVED** [review]

**Comment by MrDoomBringer** [comment]
image
image

**Comment by ExcessiveUseOfCobblestone** [comment]
doesn't the ripley let you hold several crates at once? I guess that's fine since it's less clunky, but it will be fairly inconsistent with how other objects interact.

**Comment by Bawhoppen** [comment]
@XDTM I gave plenty of legitimate reasons...
I was just offering the public a chance to trust my judgment.

That's called _DEMOCRACY._

What's the issue??? Do you hate _DEMOCRACY_ or something???

**Comment by optimumtact** [comment]
Baw has received a 7 day ban for inciting a downvote mob.

**Comment by MrDoomBringer** [comment]
@optimumtact Please refrain from posting off-topic comments on github PR threads. 

In all seriousness though, another concern is that QOL changes don't really seem to ever add much to the game, and almost always seem to actually take things away from and erode the overall game experience over time.
 
While one QOL change is no big deal, every little bit adds up to remove from some of the legitimate gameplay that should is worth experiencing through. It's powercreep-esque in nature, so as we make more and more changes like this, we become open to even more egregious ones as time goes on.
 
To illustrate this, I'll give some totally realistic QOL changes people might want to try and make, so you'll hopefully see how they will eventually end up watering the game down.
 
For example:
-Engine starts set up [it just takes a couple minutes to do so and is a routine boring task, it's just QOL to save that bit of time for everyone]
-Solars start wired [it's super easy to wire solars, why even have it be needed at all?]
-Security starts with HUDglasses and other gear on [I mean... they have to put it on ANYWAYS, why not save the few seconds?]
-Engineers start with insulated gloves on [they get them anyways, might as well start with them on]
-Science starts with first research nodes already done [oh, this just saves an extra 30 seconds or however long for the scientist... they have to do it every time anyways]
-Xenobio slimes start ready to burst [having to feed them is the same every time, let's just skip the process and save everyone time]
-Captain starts with nuke disk [it's their job to hold onto it anyways after all]
-HOP starts with all access [oh they can just give it to themselves regardless, just skip that step]
-Engineers know door wires at the start [well they have the tools to find out, no need to waste that extra time]
-Cargo crates start loaded on the cargo shuttle [it's boring to move the crates around and push them onto the shuttle every single round- we can just skip past that pointless effort]
-Captain starts with all comms enabled [they can turn them on anyways, right?]
 
This then adds up into even bigger effects over time as these changes become normalized...
 
-Repairing/disabling cameras just requires one click, rather than adjusting wires [if you have the tools, why need to fiddle around with the wires?]
-Shuttle call is now 5 minutes by default [people just do code red anyways, this saves that step]
-First research nodes are already completed [this'll save a few seconds that we have to do every time anyways]
-Cargo spare crates are removed and cargo just starts with more credits [they were already there to be shipped, why even have them?]
-Paramedic starts with teleporter access [they just ask HOP every time anyways... better streamline the process]
-Robotics starts with pre-assembled borg shell [oh it's hardly any effort to make one... will just make it easier to borg the first person to show up]
-Xenobio has slimes already on the next level up [there was no point to the original process of feeding a grey and waiting for it to burst anyways]
-Atmos starts fully upgraded with the volumetric pumps and manual valves and things [they have these pipes right away, this will just skip past them wasting the time installing them]
-Brig starts out with barriers and portable flashers pre-deployed [wardens do this anyways, we can just make it easier on them]
-Suit sensors start enabled [just a couple clicks, people should have them on anyways]
 
And eventually, this leads up to previously unthinkable things...
 
-Hacking a door doesn't have wires anymore, just does it automatically [as long as you have the tools, why fiddle around with door wires either since you can test them all anyways? just have a button that raises bolts, cuts power, and then also shocks you if you don't have gloves]
-Warops automatically distributes all access and opens armory [the crew do this anyways, just automate it for quality of life!]
-Press R to reload gun from backpack [shuffling around the inventory is annoying! we can just have it auto-reload by pressing a button, since you can just do it anyways]
-Armory, brig, bridge have turrets pre-set up [plenty of people make them, we can save them the hassle of all those bunch of steps]
-Paramedic starts with hand tele in their pocket [they'll just go get it anyways, will just save a minute with this]
-PDA light starts on [doesn't cost anyone anything! they can just turn it on anyways]
-Doctor surgery skill starts maxed [if they're able to practice for free on bodies at the start, why not just skip that?]
-Auto-surgery option [surgery is totally just a bunch of routine steps, you can freely see them on the wiki or operating console- easier for everyone to just have a hud button to press that automates it]
-Everyone starts with internals already on [oh, we all start with free internals in our bag... might as well have them on already]
 
I hope you get my point. I'm sure there are a ton of other great examples I couldn't think of, and while some of the ones I did list might not be an actual concern, I think this should hopefully show what I am worried about.

**Comment by optimumtact** [comment]
I don't think there's any reason to allow pushing multiple items at once, and it leads to fridge logic where you can push like ten crates but can only drag one

**Comment by Bawhoppen** [comment]
Sorry everyone. For some reason I thought I was being funny, but it all just came off very poorly.

**Comment by Dennok** [comment]
Why i see it like updated?

**Pull request closed without merging**